internal/api/shared/migrate/migrate.go

```diff
@@ -190,3 +194,126 @@ func (m *mig) init() error {
 	m.BuildMigrationSchemaString()
 	return nil
 }
+
+/**
+ * This function addresses an issue we have with Grafana datamodel when it comes to migrating dashboards to Perses: When
+ * a row is expanded in Grafana, its children panels are moved up in the main panels list, thus become siblings of the row.
+ * When it comes to Perses migration we need to recompose the parent->children relationships. However in its current state
+ * the CUE language doesn't permit us to achieve this recomposition, hence this processing in the backend code.
+ *
+ * So what this function does is basically the following: whenever such pattern is encountered in the panels list:
+ * ...
+ * row1,
+ * panelA,
+ * panelB,
+ * panelC,
+ * row2,
+ * ...
+ * the objects gets rearranged like:
+ * ...
+ * row1: {
+ *   panelA,
+ *   panelB,
+ *   panelC,
+ * },
+ * row2,
+ * ...
+ */
+func rearrangePanelsWithinExpandedRows(grafanaDashboardRaw json.RawMessage) json.RawMessage {
+	// unmarshall the received raw Json
+	var grafanaDashboard map[string]any
+	if err := json.Unmarshal(grafanaDashboardRaw, &grafanaDashboard); err != nil {
+		logrus.WithError(err).Error("Unable to unmarshall the received Grafana dashboard")
+		return grafanaDashboardRaw
+	}
+
+	// retrieve the panels
+	if _, found := grafanaDashboard["panels"]; !found {
+		logrus.Error(errors.New("expected `panels` field not found"))
+		return grafanaDashboardRaw
+	}
+	panels, ok := grafanaDashboard["panels"].([]any)
+	if !ok {
+		logrus.Error(errors.New("failed to assert `panels` to array of any"))
+		return grafanaDashboardRaw
+	}
+
+	// iterate over the panels & achieve recomposition of parent->children relationship when needed
+	var newPanelList []map[string]any
+	var parentRow map[string]any
+	for _, panelAsAny := range panels {
+		panel, ok := panelAsAny.(map[string]any)
+		if !ok {
+			logrus.Error(errors.New("failed to assert current panel to map of any"))
+			return grafanaDashboardRaw
+		}
+
+		if panel["type"] == "row" {
+			if _, found := panel["collapsed"]; !found {
+				logrus.Error(errors.New("expected attribute `collapsed` not found in row"))
+				return grafanaDashboardRaw
+			}
+			collapsed, ok := panel["collapsed"].(bool)
+			if !ok {
+				logrus.Error(errors.New("failed to assert the row's `collapsed` field to bool"))
+				return grafanaDashboardRaw
+			}
+
+			if collapsed && parentRow == nil {
```

Comment:
it would also help to add some example about the situation in a condition or in another. It would help to understand why we are doing the different operations in each condition.

[Here](https://github.com/prometheus/prometheus/blob/main/web/ui/module/codemirror-promql/src/complete/hybrid.ts#L284-L311
) is an example of an algo with a lot of condition where you have in almost every condition an example about the situation and why we are doing that.